base/array.jl

```diff
@@ -199,35 +199,99 @@ convert{T,n,S}(::Type{Array{T,n}}, x::AbstractArray{S,n}) = copy!(Array(T, size(
 
 promote_rule{T,n,S}(::Type{Array{T,n}}, ::Type{Array{S,n}}) = Array{promote_type(T,S),n}
 
+## copying iterators to containers
+
+# make a collection similar to `c` and appropriate for collecting `itr`
+_similar_for(c::AbstractArray, itr, T, ::SizeUnknown) = similar(c, T, 0)
```

Comment:
Maybe switch the order of `T` and `itr`, since `itr` might specify the size? (To make it more like `similar`.)